src/jvm/test/org/jetbrains/kotlinx/lincheck_test/util/TestUtils.kt

```diff
@@ -55,53 +54,114 @@ private fun compareAndOverwrite(expectedOutputFilePrefix: String, actualOutput:
+                jdkVersion = testJdkVersion,
+                inTraceDebuggerMode = isInTraceDebuggerMode,
+            )
+            val overwriteOutputFile = getExpectedLogFileFromSources(overwriteOutputFileName)
+            overwriteOutputFile.writeText(actualOutput)
+            return
         }
-    }
-}
 
-// To prevent file duplication this function finds the file to compare the results to.
-// With preference for jdk: 17, 15, 13, 11, 8 (trace mode) and 17, 15, 13, 11, 8 (non trace mode)
-// Search starts at current jdk level
-// For instance we are running tests for jdk 11 (trace mode), we will check file existence in the following order:
-// 11 (trace), 8 (trace), 11 (non-trace) and 8 (non-trace).
-private fun getFileToCompareTo(expectedOutputFilePrefix: String): String {
-    // If in trace mode first check if a trace debugger file can be found
-    if (isInTraceDebuggerMode) {
-        for (i in testJdkVersion.ordinal downTo 0) {
-            val jdkVersion = TestJdkVersion.entries[i]
-            val fileName = getFileNameFor(expectedOutputFilePrefix, jdkVersion, true)
-            if (logFileFromResourcesExists(fileName)) return fileName
+        if (expectedOutputFile == null) {
+            error(
+                """
+                No file exists yet for the test $expectedOutputFilePrefix. 
+                    JDK version = $testJdkVersion;
+                    Trace debugger mode = $isInTraceDebuggerMode.
+                """
+                .trimIndent()
+            )
         }
+
+        assertEquals(expectedOutput, actualOutput)
     }
-    // Check lower sdks in non-trace mode
-    for (i in testJdkVersion.ordinal downTo 0) {
-        val jdkVersion = TestJdkVersion.entries[i]
-        val fileName = getFileNameFor(expectedOutputFilePrefix, jdkVersion, false)
-        if (logFileFromResourcesExists(fileName)) return fileName
+}
+
+/* To prevent file duplication, this function finds the file to compare the results to.
+ * It first tries to find a jdk- and mode-specific file.
+ * If there is no such file, it then looks up for a default file
+ * (i.e., file for a default JDK version and non-trace mode).
+ *
+ * For instance, we are running tests for jdk 11 (trace mode),
+ * we will check file existence in the following order:
+ *   fn_trace_debugger_jdk11.txt
+ *   fn_jdk11.txt
+ *   fn_trace_debugger.txt
+ *   fn.txt
+ *
+ * Returns null if none of these files exists.
+ */
+private fun getExpectedLogFile(expectedOutputFilePrefix: String): File? {
+    data class TestFileConfiguration(
+        val jdkVersion: TestJdkVersion,
+        val inTraceDebuggerMode: Boolean
+    )
+
+    val testConfigurations = listOfNotNull(
+        // first try to pick the most specific file if it exists
+        TestFileConfiguration(
+            jdkVersion = testJdkVersion,
+            inTraceDebuggerMode = isInTraceDebuggerMode
+        ),
+
+        // next, try to pick the jdk-specific file
+        TestFileConfiguration(
```

Comment:
Suggestion: this can also be wrapped in `if (isInTraceDebuggerMode)` and the bottom two in `if (testJdkVersion != DEFAULT_TEST_JDK_VERSION)`

Reply:
I'll keep it as is. The logic is already complicated enough, I would prefer to keep the number of `if`-s minimal.